addons/purchase_stock/models/product.py

```diff
@@ -69,3 +69,42 @@ def _get_quantity_in_progress(self, location_ids=False, warehouse_ids=False):
             qty_by_product_location[(product.id, location.id)] += product_qty
             qty_by_product_wh[(product.id, location.warehouse_id.id)] += product_qty
         return qty_by_product_location, qty_by_product_wh
+
+
+class SupplierInfo(models.Model):
+    _inherit = 'product.supplierinfo'
+
+    last_purchase_date = fields.Date('Last Purchase', compute='_compute_last_purchase_date')
+    show_set_supplier_button = fields.Boolean(
+        'Show Set Supplier Button', compute='_compute_show_set_supplier_button')
+
+    def _compute_last_purchase_date(self):
+        self.last_purchase_date = False
+        for supplier in self:
+            purchase = self.env['purchase.order'].search([
+                ('state', 'in', ('purchase', 'done')),
+                ('order_line.product_id', 'in', self.product_tmpl_id.product_variant_ids.ids),
+                ('partner_id', '=', supplier.name.id),
+            ], order='date_order', limit=1)
+            if purchase:
+                supplier.last_purchase_date = purchase.date_order
+
+    def _compute_show_set_supplier_button(self):
+        self.show_set_supplier_button = True
+        orderpoint_id = self.env.context.get('default_orderpoint_id')
+        orderpoint = self.env['stock.warehouse.orderpoint'].browse(orderpoint_id)
+        if orderpoint_id:
+            self.filtered(
+                lambda s: s.id == orderpoint.supplier_id.id
+            ).show_set_supplier_button = False
+
+    def action_set_supplier(self):
```

Comment:
ensure_one and remove the loop. No sense to set several supplier to the same orderpoint, right ?